libraries/chain/smt_evaluator.cpp

```diff
@@ -300,5 +307,31 @@ void smt_set_runtime_parameters_evaluator::do_apply( const smt_set_runtime_param
    });
 }
 
+void smt_contribute_evaluator::do_apply( const smt_contribute_operation& o )
+{
+   try
+   {
+      FC_ASSERT( _db.has_hardfork( STEEM_SMT_HARDFORK ), "SMT functionality not enabled until hardfork ${hf}", ("hf", STEEM_SMT_HARDFORK) );
+
+      const smt_token_object* token = util::smt::find_token( _db, o.symbol );
+
+      FC_ASSERT( token != nullptr, "Cannot contribute to an unknown SMT" );
+      FC_ASSERT( token->phase >= smt_phase::contribution_begin_time_completed, "SMT has yet to enter the contribution phase" );
+      FC_ASSERT( token->phase < smt_phase::contribution_end_time_completed, "SMT is no longer in the contribution phase" );
+      FC_ASSERT( _db.get_balance( o.contributor, o.contribution.symbol ) >= o.contribution, "Account does not have sufficient funds for contribution" );
+
+      _db.adjust_balance( o.contributor, -o.contribution );
+
+      _db.create< smt_contribution_object >( [&] ( smt_contribution_object& obj )
+      {
+         obj.contributor = o.contributor;
+         obj.symbol = o.symbol;
+         obj.contribution_id = o.contribution_id;
```

Comment:
There is no check for uniqueness of contribution id other than letting the creation of the object fail and a horrible exception thrown to the user.